### ISSUE TITLE
[Question] Tileset drawing

### ISSUE THREAD

**Issue by linde12** [ggez/ggez#389]
Hey! Just found this lib and was amazed at how easy it was to get started compared to the alternatives. Now when i've got the basics down i'm trying to load my tileset in a basic way but i'm confused with how i should render the individual tiles.

In the code below i tried try to specify the source rect in pixels, but then saw in the docs that the the value of the x and y could be max 1.0. I'm not sure if there is another way to do this? In C++ with SDL2 i simply passed x, y, width and height (all in pixels). Is there a better way of doing this? I get the feeling that this is not the proper way to do it.

Just to clarify, i have a loaded tileset in an `ggez::graphics::Image` and i want to draw the individual "tiles" with ggez.

```
    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx);
        graphics::draw_ex(ctx, &self.tileset, DrawParam {
            src: graphics::Rect::new(0.0, 0.0, 1.0 / 12.0, 1.0/12.0),
            dest: graphics::Point2::new(self.pos_x, self.pos_y),
            .. Default::default()
        }).expect("cannot draw tile");

        graphics::present(ctx);
        Ok(())
    }
```

**Comment by icefoxen** [comment]
Hi!  Glad you like it.

The reason ggez uses `[0...1]` is basically because graphics cards generally don't know or care about pixels these days.  ggez's default setup arranges it so that 1 texel in your texture (ie your `graphics::Image`) lines up with 1 pixel on your screen, but it's entirely possible to zoom, warp or twist the image or ggez's projection to make it whatever you want.  If you zoom so that your 32x32 texel image fills 128x14 pixels on the screen, and tell it to draw an 8x8 segment of it, what SHOULD happen?

(Ok, it should take the 8x8 texel segment and then apply the transformation, but still.)

So, what you are doing is just fine.  We might have to add a better explanation of this to the docs.  We may want to consider changing the API so that images are addressed in texels... but that would make life harder for other situations, such as UV'ing meshes or doing LOD'ing (if we ever have anyone who cares about LOD'ing?).

**Comment by linde12** [comment]
I see! Thanks for the explaination, i guess i'm just not used to it yet. Being able to choose would be great i think.

Thanks again!

**Comment by beatgammit** [comment]
I got tripped up by this as well, but I ended up not fixing the docs because I figured I was just bad at reading comprehension (the docs *clearly* stated max of 1.0).

I think a builder would be awesome. It would be especially awesome if we could supply logical units, like `get_tile(x: i32, y: i32, tile_width: i32, tile_height: i32)`, which would internally do something like:

    Rect::new(
        (x as f32) / ((image_width / tile_width) as f32),
        (y as f32) / ((image_height / tile_height) as f32),
        1.0 / (tile_height as f32),
        1.0 / (tile_height as f32),
    )

**Comment by icefoxen** [comment]
Yeah, the issue is that a `DrawParam` has no idea what an `Image`'s size is, which means we'd have to do something silly like

```rust
graphics::draw(ctx, image, DrawParam::new(image.get_dimensions())
                                               .src_pixels(x, y, w, h))?;
```

...or such.  Which then doesn't make much sense when trying to draw, say, a `Mesh` or `Text`.

The sanest thing I can think of so far is having a flag on `Image`, or newtype wrapper around `Image`, which sets the coordinate system.  So one way, `src_rect`, `offset`, maybe `scale`, will all be in pixels, the other way they are all in multiples of the image size.

**Comment by icefoxen** [comment]
That would be nice, since tile maps is one of the main use cases of `SpriteBatch`.

**Comment by EmmChriss** [comment]
What about the DrawParam containing either texcoords or texels in a wrapper enum, and adjusting the behaviour in `.draw` and `.draw_ex` accordingly.
This should of course panic if the texels are out of bound

**Comment by icefoxen** [comment]
https://github.com/ggez/ggez-goodies/blob/devel/src/tilemap.rs

This is a bit crude but it should be released along with 0.5.
